client/extensions/woocommerce/woocommerce-services/state/carrier-accounts/reducer.js

```diff
@@ -0,0 +1,145 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import { isEmpty } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import {
+	WOOCOMMERCE_SERVICES_CARRIER_ACCOUNTS_DISCONNECT_CARRIER,
+	WOOCOMMERCE_SERVICES_CARRIER_ACCOUNTS_ENABLE_CANCEL_CONNECTION_DIALOG,
+	WOOCOMMERCE_SERVICES_CARRIER_ACCOUNTS_ENABLE_DISCONNECT_DIALOG,
+	WOOCOMMERCE_SERVICES_CARRIER_ACCOUNTS_SUBMIT_SETTINGS,
+	WOOCOMMERCE_SERVICES_CARRIER_ACCOUNTS_UPDATE_SETTINGS,
+	WOOCOMMERCE_SERVICES_CARRIER_ACCOUNTS_TOGGLE_SHOW_UPS_INVOICE_FIELDS,
+} from '../action-types';
+
+export const initialState = {
+	modalErrors: {},
+	pristine: true,
+};
+
+const reducers = {};
+
+reducers[ WOOCOMMERCE_SERVICES_CARRIER_ACCOUNTS_SUBMIT_SETTINGS ] = ( state, { carrier } ) => {
+	const settings = state[ carrier ].settings;
+	const { fieldErrors } = settings;
+
+	if ( ! isEmpty( fieldErrors ) ) {
+		return state;
+	}
+
+	const newState = {
+		...state,
+		[ carrier ]: { settings: { ...settings, fieldErrors } }
+	};
+
+	return newState;
+};
+
+reducers[ WOOCOMMERCE_SERVICES_CARRIER_ACCOUNTS_UPDATE_SETTINGS ] = ( state, { carrier, fieldName, newValue } ) => {
+	const settings = state[ carrier ].settings;
+	const { ignoreValidation, values } = settings;
+
+	if ( newValue.target ) {
+		switch( newValue.target.type ) {
+			case 'checkbox':
+				newValue = newValue.target.checked;
+				break;
+			default:
+				newValue = newValue.target.value;
+		}
+	}
+
+	const newState = {
+		...state,
+		[ carrier ]: {
+			settings: {
+				...settings,
+				values: {
+					...values,
+					[ fieldName ]: newValue
+				}
+			}
+		},
+	};
+
+	if ( ignoreValidation ) {
+		newState[carrier].settings.ignoreValidation = {
+			...ignoreValidation,
+			[ fieldName ]: false,
+		};
+	}
+
+	return newState;
+};
+
+reducers[ WOOCOMMERCE_SERVICES_CARRIER_ACCOUNTS_TOGGLE_SHOW_UPS_INVOICE_FIELDS ] = ( state, { carrier } ) => {
+	const settings = state[ carrier ].settings;
+
+	const newState = {
+		...state,
+		[ carrier ]: {
```

Comment:
If this reducer (and its action) is strictly for `UPS`, should we remove `carrier` from the parameter? and have it as 
```
UPS: {
  ...state,
  settings: {...}
}
```

Otherwise, we could have an action `WOOCOMMERCE_SERVICES_CARRIER_ACCOUNTS_TOGGLE_SHOW_UPS_INVOICE_FIELDS` for `fedex` and it will have `showUPSInvoiceFields` as a settings state.

Reply:
Hmm, yeah your right, that action and variable names are not right. I'll change them to make them carrier agnostic. I wanted to make this code work for any carrier we may support in the future.